### ISSUE TITLE
Plugin system

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fable-compiler/Fable#2142]
See #2066, although that PR is going to be superseded by a new one targeting Nagareyama.

**Comment by Shmew** [comment]
I've been talking with @Zaid-Ajaj about the plugin system and had some questions/ideas:

* Is it possible to transform a declaration from the call site?
* Is it possible to determine if some item has already been visited and thus transformed?

To elaborate on why I'm asking these, the way it's looking to be shaping up now users will need to put an attribute on every function component they define. If they fail to do this, it will *not* fail and may cause issues/unexpected behavior at runtime. The idea that I had is that since we know that *every* instance of a function that is *not* inlined and returns `ReactElement` should be a component, I would think we can make a more broad solution that helps prevent this situation.

The idea is that instead of defining every function component with an attribute we instead apply a React *entry* attribute. Since React itself is a tree everything that is used by the application will go through this entry point. Since this is the case, it would be really beneficial if we could apply the transformations this way. 

It does raise some questions for some valid (if unlikely) cases where a user may want to consume Fable from Javascript, as then the author would need to apply the original attribute system... which then goes into the second question about determining if something has already been transformed.

Another case regarding determining if something has been visited is cases where we might be able to apply optimizations such as take this for example:

```fsharp
let someChild (compute: int -> int) =
    ...

let counterDisplay (someNumber: int) (compute: int -> int) =
    let displayedCount = compute someNumber
    
    React.fragment [
        Html.h1 displayedCount
        someChild compute
    ]
    

[<ReactEntry>]
let counter (initialCount: int) =
    let (count, setCount) = React.useState initialCount

    let compute x = count - initialCount + x

    Html.div [
        Html.button [
            prop.style [ style.marginRight 5 ]
            prop.onClick (fun _ -> setCount(count + 1))
            prop.text "Increment"
        ]
        Html.button [
            prop.style [ style.marginLeft 5 ]
            prop.onClick (fun _ -> setCount(count - 1))
            prop.text "Decrement"
        ]
        counterDisplay count compute
    ]
```

It would be nice if using our plugin we could automatically apply `React.useCallback` to the `compute` function, but *only* if it's passed down to another React component and itself is not already explicitly wrapped with `React.useCallback`. However, once it gets to the `counterDisplay` component it should *not* be transformed even though it's then passed further down to `someChild`.

I don't know if this is really possible currently or future state, but it would really open the doors to writing very *native* feeling F# code while gaining a lot of optimizations under the hood that would normally require a decent understanding of how React works.

**Comment by Zaid-Ajaj** [comment]
> Is it possible to transform a declaration from the call site?

@Shmew Yeah we already do this in the plugins

> Is it possible to determine if some item has already been visited and thus transformed?

You could emit harmless metadata and detect it later on. However, I feel like this "application-wide" transform is really like fork of the compiler, tailored for React. Even if I like the idea, it is simply not possible to implement with the capacity we have and it really really complicated! 

I don't think that forgetting the attribute would warrant such an endeavour. That said, I think we can _relatively_ easily build an **analyzer** to tell users about things like:
 - Missing `[<ReactComponent>]` when hooks are used
 - Not calling `React.useCallback` to stabilize a reference when a callback is used as an input property
 - Not providing proper values in the dependency arrays when some input prop is clearly a dependency

What I am saying is, better tell users what they might be missing instead of rewriting their code to do damage control

**Comment by alfonsogarciacaro** [comment]
Those are very interesting ideas @Shmew, thanks for sharing them! Although I'm with @Zaid-Ajaj here in that trying to do too much with the plugins may go out of control quickly. It's true that you won't get a compiler warning if you forget the attribute (in ReasonReact they use attributes as well, but they use JSX so I guess the call site will fail if the function is not an actual component), but converting all functions into components may be a bit too aggressive. In my projects I usually have many helper function components per file but only one component.

> You could emit harmless metadata and detect it later on.

Actually, right no it's not possible (or at least not easy) to put metadata in Fable AST. I was thinking about it but wasn't sure how to implement it. Probably we could recycle the `TypeCast` element to minimize code changes and add an optional `tag` (string or obj?) element. Did you have something in mind?

**Comment by Zaid-Ajaj** [comment]
> Did you have something in mind?

I did have something in mind but probably not a very good idea: emitting magic string or numbers which we can detect later. Now when read this sentence again, it does sounds like a really bad idea 😅 but in any case, I think this problem here is something for the analyzers to help rather than plugins. That said, even without those, I can live without either. The overhead of adding such attribute is really minimal

**Comment by Shmew** [comment]
Yeah I'd really like analyzers if they just worked in every major IDE 😢. The idea spawned not only from people needing to remember the attributes and the surprises that can pop up later because of it although, it's not like this isn't the case right now so it's not a big deal. The more "big" idea was being able to write code that looks even more similar to native F# code while still getting the proper benefits of React (in this case, but this idea isn't limited to that of course). Which can be a double edged sword depending on if you're someone that does want to actually learn how to use things like React.

The AST itself isn't the output so is there a reason to use more complex data structures instead of magic strings? If the plugin is able to have a *unique* instance of the tag/metadata/whatever (I don't actually know how this part works) then it seems like it would make it really easy to extend extra functionality in a plugin. Like the example above with `React.useCallback` being able to have plugin use something like this for the tag:

```fsharp
type Transform =
    | UseCallback
    | UseMemo
    | FunctionComponent

type Transforms = Set<Transform>

type MyAttribute () =
    inherit MemberDeclarationPluginAttribute<Transforms>()
    override _.FableMinimumVersion = "3.0"

    override _.TransformCall(logger, memb, expr) = ...
```
